standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/DirectSqlUpdatePart.java

```diff
@@ -181,7 +182,7 @@ private void populateInsertUpdateMap(Map<PartitionInfo, ColumnStatistics> statsP
             e -> e.partitionId).collect(Collectors.toList()
     );
 
-    prefix.append("select \"PART_ID\", \"COLUMN_NAME\" from \"PART_COL_STATS\" WHERE ");
+    prefix.append("select \"PART_ID\", \"COLUMN_NAME\", \"ENGINE\" from \"PART_COL_STATS\" WHERE ");
```

Comment:
Why do we need `ENGINE ` field here?
Isn't the `PART_ID ` field enough to identify the specific partition?

Reply:
It could be possible,
https://github.com/apache/hive/blob/master/standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java#L10130
in the above example, we use different engines to fetch the statistics with the same partition and column.

Reply:
Make sense.  if so, Shouldn't we add the `ENGNE `field in the `WHERE `sub-statement?
Current direct SQL query:
`select "PART_ID", "COLUMN_NAME" from "PART_COL_STATS" WHERE ("PART_ID" in (141))`

Maybe can be changed to:
`select "PART_ID", "COLUMN_NAME" from "PART_COL_STATS" WHERE ("PART_ID" in (141)) and ENGINE = hive`

BTW, we also have another field `CAT_NAME` in `PART_COL_STATS` to differentiate column stas between multi catalog. Should we also consider it here?

Reply:
Morning @zhangbutao!
In this method we want to get the insert or the update statistics from `Map<PartitionInfo, ColumnStatistics> statsPartInfoMap`, there is no guarantee that all of the `statsPartInfoMap` are for the engine hive or the same engine, so PartColNameInfo needs to feed with the engine info when compared with the stats in `statsPartInfoMap`.

> BTW, we also have another field CAT_NAME in PART_COL_STATS to differentiate column stas between multi catalog. Should we also consider it here?

I think we don't need to, the `PART_ID` here has the same effect for clarifying the catalog.

Reply:
Thanks for explanation!

> there is no guarantee that all of the statsPartInfoMap are for the engine hive or the same engine,

Is this code snippet only for `insert` statement?

If so, the one `insert` statement is must from the one same `ENGINE` (like hive or other compute engine), so the `statsPartInfoMap` will have the same `ENGINE `value.

Reply:
> Is this code snippet only for insert statement?

No, as you see, this PR aims for optimizing the partitioned table rename. Even the `statsPartInfoMap` has the same `ENGINE` value, we must be careful not to override the stats(for other engines) for the same partition column on the database.

Reply:
I did some code debug. Only found that `insert` into statement gone through this code block(method `populateInsertUpdateMap`), but `alter table rename` not. Maybe i missed some configuration?

Since `statsPartInfoMap` has the same `ENGINE `value, why not we use this direct sql to filter the PART_ID based on the `ENGINE` (e.g. `hive`)? 
`select "PART_ID", "COLUMN_NAME" from "PART_COL_STATS" WHERE ("PART_ID" in (141)) and ENGINE = hive`

Reply:
@zhangbutao, you need to apply this change to see `populateInsertUpdateMap` is invoked by the rename operation, there is a batch rename for stats.
In my idea, there is no document that all the stats in `statsPartInfoMap` must be the same engine, so we cannot simply using the `ENGINE = hive` for all stats, besides `ENGINE = hive` in the filter is almost the same as in `select \"ENGINE\"` when there has limited numbers of engines on this partition.

Reply:
ok,i didn't apply your change to do test. No need to consideration about this comment, i just want to explore more details about stats usage in Hive.
Thanks. 😃